_posts/2018-01-25-totw-3.md

```diff
@@ -0,0 +1,94 @@
+---
+title: "Tip of the Week #3: String Concatenation and operator+ vs. StrCat()"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/3
+type: markdown
+order: "003"
+---
+
+Originally published as totw/3 on 2012-05-11
+
+Updated 2017-09-18; revised 2018-01-22
+
+Users are often surprised when a reviewer says, "Don't use the string
+concatenation operator, it's not that efficient." How can it be that
+`string::operator+` is inefficient? Isn't it hard to get that wrong?
+
+It turns out, such inefficiency isn’t clear cut. These two snippets have
+close to the same execution time, in practice:
+
+```cpp
+string foo = LongString1();
+string bar = LongString2();
+string foobar = foo + bar;
+
+string foo = LongString1();
+string bar = LongString2();
+string foobar = absl::StrCat(foo, bar);
+```
+
+However, the same is not true for these two snippets:
+
+```cpp
+string foo = LongString1();
+string bar = LongString2();
+string baz = LongString3();
+string foobar = foo + bar + baz;
+
+string foo = LongString1();
+string bar = LongString2();
+string baz = LongString3();
+string foobar = absl::StrCat(foo, bar, baz);
+```
+
+The reason these two cases differ can be understood when we pick apart what is
+happening in the expression `foo + bar + baz`. Since there are no overloads for
+three-argument operators in C++, this operation is necessarily going to make two
+calls to `string::operator+`. And between those two calls, the operation will
+construct (and store) a temporary string. So `string foobar = foo + bar + baz`
+is really equivalent to:
+
+```cpp
+string temp = foo + bar;
+string foobar = std::move(temp) + baz;
+```
+
+Specifically, note that the contents of `foo` and `bar` must be copied to a
+temporary location before they are placed within `foobar`. (For more on
+`std::move`, see
+[Tip of the Week #77: Temporaries, moves, and copies](/tips/77).)
+
+C++11 at least allows the second concatenation to happen without creating a new
+string object: `std::move(temp) + baz` is equivalent to
+`std::move(temp.append(baz))`. However, it's possible that the buffer initially
```

Comment:
Parens in the wrong place? I think this wants to be std::move(temp).append(baz) ?

I'm actually slightly unsure what this is aiming for all of a sudden.